relayer/chain/ethereum/listener.go

```diff
@@ -151,34 +163,54 @@ func (li *Listener) queryEvents(ctx context.Context, contract *outbound.Contract
 			}
 			break
 		}
+		events = append(events, &iter.Event.Raw)
+	}
+	return events, nil
+}
+
+func (li *Listener) queryIncentivizedEvents(contract *outbound.IncentivizedOutboundChannel, options *bind.FilterOpts) ([]*etypes.Log, error) {
+	var events []*etypes.Log
 
-		events = append(events, iter.Event)
+	iter, err := contract.FilterMessage(options)
+	if err != nil {
+		return nil, err
 	}
 
+	for {
+		more := iter.Next()
+		if !more {
+			err = iter.Error()
+			if err != nil {
+				return nil, err
```

Comment:
What's the reason to not return the already-processed events in case of error?

Reply:
When relaying messages, we're following an all-or-nothing approach. The relayer either successfully relays all the messages, or it fails-fast and restarts.

This simplifies error-handling. We can't easily just forward a subset of the messages/events because the message nonce's need to match up with what the Substrate inbound channel expects.